dashboard/dashboard.go

```diff
@@ -0,0 +1,304 @@
+}
+
+type data struct {
+	Time  time.Time `json:"time,omitempty"`
+	Value float64   `json:"value,omitempty"`
+}
+
+type status struct {
+	Peers int `json:"peers,omitempty"`
+	Block int `json:"block,omitempty"`
+}
+
+// New creates a new dashboard instance with the given configuration
+func New(config *Config) (*dashboard, error) {
+	//log.Trace("NewDashboard() called")
+
+	dashboard := &dashboard{
+		config:  config,
+		Metrics: &metricSamples{},
+		quit:    make(chan struct{}),
+	}
+
+	if config.Assets == "" {
+		tmpl, err := Asset("dashboard.html")
+		if err != nil {
+			return nil, err
+		}
+
+		website := new(bytes.Buffer)
+		// set the sample limits for the client
+		if err = template.Must(template.New("").Parse(string(tmpl))).Execute(website, map[string]interface{}{
+			"processorSampleLimit": processorSampleLimit,
+			"memorySampleLimit":    memorySampleLimit,
+		}); err != nil {
+			log.Crit("Failed to render the dashboard template", "err", err)
+		}
+
+		dashboard.index = website.Bytes()
+		return dashboard, nil
+	}
+
+	//TODO (kurkomisi): case DashboardAssetsFlag is set
+	//dashboard.index = ioutil.ReadFile()
+	return dashboard, nil
+}
+
+// Protocols is a meaningless implementation of node.Service
+func (db *dashboard) Protocols() []p2p.Protocol { return nil }
+
+// APIs is a meaningless implementation of node.Service
+func (db *dashboard) APIs() []rpc.API { return nil }
+
+// Start implements node.Service, starting the data collection thread and the listening server of the dashboard
+func (db *dashboard) Start(server *p2p.Server) error {
+	//log.Trace("Start() called", "config", db.config)
+
+	go db.collectData()
+
+	http.HandleFunc("/", db.webHandler)
+	http.Handle("/api", websocket.Handler(db.apiHandler))
+
+	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", db.config.Host, db.config.Port))
+	if err != nil {
+		return err
+	}
+	db.listener = listener
+
+	go func() {
+		//log.Trace("Starting server...")
+
+		if err := http.Serve(listener, nil); err != nil {
+			log.Warn("Server failed", "err", err)
+		}
+	}()
+
+	return nil
+}
+
+// Stop implements node.Service, stopping the data collection thread and the connection listener of the dashboard
+func (db *dashboard) Stop() error {
+	//log.Trace("Terminating dashboard...")
+
+	var err error
+
+	// Close the connection listener
+	db.lock.Lock()
+	if err = db.listener.Close(); err != nil {
+		log.Warn("Failed to close listener", "err", err)
+	}
+
+	// Notifies collectData and apiHandler
+	close(db.quit)
+
+	for _, c := range db.conns {
+		if err := c.conn.Close(); err != nil {
+			c.logger.Warn("Failed to close connection", "err", err)
+		}
+	}
+	db.conns = db.conns[:0]
+	db.lock.Unlock()
+
+	return err
+}
+
+// webHandler handles all non-api requests, simply flattening and returning the dashboard website.
+func (db *dashboard) webHandler(w http.ResponseWriter, r *http.Request) {
+	//log.Trace("webHandler() called")
+
+	//TODO (kurkomisi): not only index
+	w.Write(db.index)
```

Comment:
Instead of returning `db.index` here, you'll need to parse the r.URL.Path (or something like that) to retrieve the actual asset the user requested. If that's `/` or `/index.html` you can return the current index, otherwise pull in the requested asset from the embedded data `Asset("user.request")` (or serve from the file system if the Asset flag was specified).